src/app/state/persist/saga.ts

```diff
@@ -0,0 +1,165 @@
+      yield* put(persistActions.setWrongPassword())
+    } else {
+      throw error
+    }
+  }
+}
+
+function* lockAsync(action: ReturnType<typeof persistActions.lockAsync>) {
+  yield* put(persistActions.resetRootState())
+  // Implies state.loading = false
+}
+
+function* eraseAsync(action: ReturnType<typeof persistActions.eraseAsync>) {
+  yield* call([localStorage, localStorage.removeItem], STORAGE_FIELD)
+  yield* put(persistActions.resetRootState())
+  // Implies state.loading = false
+}
+
+function* resetRootState(action: ReturnType<typeof persistActions.resetRootState>) {
+  const unlockedStatus = yield* select(selectUnlockedStatus)
+  // Redirect home to prevent infinite loading if user closes unpersisted wallet in another tab.
+  if (isActionSynced(action) && unlockedStatus === 'emptyUnpersisted') {
+    // Note: can only redirect in webapp, because saga runs in background page in extension.
+    // TODO: find another way if extension starts supporting multiple popups or tabs
+    // e.g. show "You closed wallet in another tab, click to go home"
+    if (runtimeIs === 'webapp') {
+      yield* call([window.location, window.location.assign], '/')
+    }
+  }
+}
+
+// Encrypting
+async function encryptState(state: RootState, keyWithSalt: KeyWithSalt): Promise<EncryptedString> {
+  const persistedRootState: PersistedRootState = {
+    theme: state.theme,
+    wallet: state.wallet,
+    network: state.network,
+  }
+  const encryptedState = await encryptWithKey(keyWithSalt, persistedRootState)
+  return encryptedState
+}
+
+async function decryptState(
+  encryptedState: EncryptedString,
+  password: string,
+): Promise<SetUnlockedRootStatePayload> {
+  const persistedRootState: PersistedRootState = await decryptWithPassword(password, encryptedState)
+  const keyWithSalt = await deriveKeyFromPassword(password)
+  const stringifiedEncryptionKey = base64andStringify(keyWithSalt)
+  return { persistedRootState, stringifiedEncryptionKey }
+}
+
+function* encryptAndPersistState(action: AnyAction) {
+  /**
+   * Latest state does not match state immediately after action was dispatched,
+   * but when it is queued by {@link watchPersistAsync}.
+   */
+  const latestState: RootState = yield* select()
+  if (latestState.persist.stringifiedEncryptionKey === 'skipped') return
+  if (action.type.startsWith('@')) return // Ignore @@INIT from redux-devtools-instrument
+  if (action.type.startsWith('&')) return // Ignore e.g. &_GET_INIT_STATE from redux-state-sync
+  if (isActionSynced(action)) return // Ignore actions synced across tabs from redux-state-sync
+
+  // Handle actions while state is locked
+  // TODO: If we used private/public key-pair we could encrypt and persist
+  // state while state is locked. But not if we support multiple profiles.
+  if (!latestState.persist.stringifiedEncryptionKey) {
+    if (walletActions.walletOpened.match(action)) {
+      throw new Error('Could not add account while state is locked.')
```

Comment:
security review: what's this for? coverage indicates we don't test this. do we catch this somewhere?

Reply:
Context: Next line skips writing on each change if wallet is locked.
For example:
- user might switch network
- it starts loading
- switch to another tab, click lock
- network info loads and dispatches networkSelected
- skip writing because we don't have the key to write
- after unlocking old network is still selected

I think it's acceptable to ignore almost all actions like this, except when silently losing an account:
- add an account from private key
- switch to another tab, click lock
- account info loads and dispatches addWallet
- throw error because we don't have the key to write

Reply:
couldn't understand that

is the network selection somehow related with unlocking?

in the second example, why wouldn't we have the key to write? is there a way to add an account from private key before setting up a password?

Reply:
> is the network selection somehow related with unlocking?

If you mean in relation to "after unlocking old network is still selected": selected network is persisted after network info loads. So normally it would keep last selected network after unlocking.

> in the second example, why wouldn't we have the key to write? is there a way to add an account from private key before setting up a password?

when user locks one tab that syncs to other tabs and it removes stringifiedEncryptionKey from state. So we can't encrypt the new account to storage

Reply:
is there a more general way to stop sagas (?) when locking?

thanks for explaining

Reply:
There might be, with `sagaMiddleware.run(rootSagas).cancel()`, but I haven't looked at actually implementing it and what the possible consequences could be.

None-general way would be to add `race({ task, cancel: cancel: take(persistActions.lockAsync) })` to most tasks, except the infinitely running ones. Or alternatively wait for most tasks to finish before locking (except infinitely running).

I rate it: too much effort

Reply:
k